Fix small Averager issues

mgiesler opened the pull request:
## Proposed changes

Make Averager moves noexcept.
This was to fix the error associated with trying to specify noexcept on the 
move constructor in FunctionOfTimeUpdater (since Averager is a member 
variable of FunctionOfTimeUpdater). The Averager was noexcept due to 
member variables of type std::deque and boost::optional, neither of which 
have noexcept moves.

### Types of changes:

- [x] Bugfix
- [ ] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

mgiesler committed:
Make Averager move noexcept

nilsdeppe commented on src/ControlSystem/Averager.cpp:
```diff
@@ -23,9 +23,22 @@ Averager<DerivOrder>::Averager(const double avg_timescale_frac,
 template <size_t DerivOrder>
 boost::optional<std::array<DataVector, DerivOrder + 1>> Averager<DerivOrder>::
 operator()(const double time) const noexcept {
+  // NOTE: here we return a copy. If this is not desired, then we need two boost
```
Since it is an array of DataVector the copy will be fairly expensive. I'd be fine with having a second variable that is the `boost::none`.

nilsdeppe commented on tests/Unit/ControlSystem/Test_Averager.cpp:
```diff
@@ -213,3 +214,24 @@ SPECTRE_TEST_CASE("Unit.ControlSystem.Averager.BadCallToAverageTime",
+  auto avg_q = averager.using_average_0th_deriv_of_q();
+  auto avg_values = averager(0.9).get();
+  // check moved values against stored values
+  auto new_averager(std::move(averager));
```
I believe another useful test to make sure things are noexcept would [be](https://en.cppreference.com/w/cpp/types/is_move_constructible) `static_assert(std::is_nothrow_move_constructible<Averager>, "");` and similar for

nilsdeppe commented on tests/Unit/ControlSystem/Test_Averager.cpp:
```diff
@@ -213,3 +214,24 @@ SPECTRE_TEST_CASE("Unit.ControlSystem.Averager.BadCallToAverageTime",
+  // check moved values against stored values
+  auto new_averager(std::move(averager));
+  CHECK(last_time == new_averager.last_time_updated());
+  CHECK(avg_time == new_averager.average_time(0.9));
```
Can you also check move assignment?

nilsdeppe commented on src/ControlSystem/Averager.hpp:
```diff
@@ -45,6 +45,15 @@ class Averager {
+  // Explicitly defined move constructor due to the fact that the std::deque
+  // move constructor is not marked noexcept
+  Averager(Averager&& A) noexcept;
+  Averager& operator=(Averager&&) noexcept;
```
I think this is missing a definition. Maybe I missed it.

nilsdeppe reviewed (changes requested):
I'm happy for you to squash the changes :)

mgiesler commented:
@nilsdeppe ok, squashed. I added the checks you suggested and I believe I have fixed the copy issue in the call operator -- but, please let me know if I am mistaken. Thanks!!

nilsdeppe commented on tests/Unit/ControlSystem/Test_Averager.cpp:
```diff
@@ -213,3 +214,36 @@ SPECTRE_TEST_CASE("Unit.ControlSystem.Averager.BadCallToAverageTime",
+  CHECK(avg_q == new_averager.using_average_0th_deriv_of_q());
+  CHECK(avg_values[0][0] == new_averager(0.9).get()[0][0]);
+  // test move assignment
+  auto new_averager2 = std::move(new_averager);
```
This still tests move construction. You want something like:

```
Averager new_averager2{};
new_averager2 = std::move(new_averager);
```

mgiesler replied:
Ahh, of course, I need to construct before assigning. Thanks for catching that.

nilsdeppe commented on src/ControlSystem/Averager.hpp:
```diff
@@ -45,6 +45,27 @@ class Averager {
+  // Explicitly defined move constructor due to the fact that the std::deque
+  // move constructor is not marked noexcept
+  Averager(Averager&& A) noexcept;
+  Averager& operator=(Averager&& A) noexcept {
```
The definition can go in the cpp file and we have been using `rhs` as the variable, not `A` (same variable change for move constructor)

nilsdeppe commented on src/ControlSystem/Averager.cpp:
```diff
@@ -23,13 +23,24 @@ Averager<DerivOrder>::Averager(const double avg_timescale_frac,
+      tau_k_(std::move(A.tau_k_)) {}
+
 template <size_t DerivOrder>
 boost::optional<std::array<DataVector, DerivOrder + 1>> Averager<DerivOrder>::
```
Can this return a `const boost::optional<...>&` now?

nilsdeppe commented on src/ControlSystem/Averager.hpp:
```diff
@@ -93,7 +114,9 @@ class Averager {
   double avg_tscale_frac_;
   bool average_0th_deriv_of_q_;
-  boost::optional<std::array<DataVector, DerivOrder + 1>> averaged_values_{};
+  boost::optional<std::array<DataVector, DerivOrder + 1>> averaged_values_;
```
Is this change (removing `{}`) actually necessary?

mgiesler replied:
Nope, that was unintentional, thanks for pointing that out.

nilsdeppe commented on src/ControlSystem/Averager.hpp:
```diff
@@ -45,6 +45,27 @@ class Averager {
+    }
+    return *this;
+  }
+  Averager(const Averager&) = delete;
```
This class used to be copyable, are you sure you don't want it to be anymore? I think the `default` copy semantics would work fine if you needed them. (Note: Not a change request, just confirming this is what you want :) )

mgiesler replied:
Yeah, I meant to do this one. We made the other components of the control system non-copyable, so I made the Averager non-copyable as well.

nilsdeppe reviewed (changes requested):
I'm happy for you to squash these last few minor things :)

mgiesler commented:
@nilsdeppe, thanks for the review -- squashed again.

kidder commented:
@wthrowe any comments on this?

wthrowe commented:
Code looks fine, but Travis OSX has a non-timeout failure that I'm not sure what to make of.

kidder commented:
os x failure looks to be from #896

nilsdeppe commented:
This just needs a rebase on develop and it should be ready to go :+1:

mgiesler commented:
Ok, rebased.
